posts/history_download/index.md

```diff
@@ -0,0 +1,316 @@
+    for message in instance.serialize()[1:]
+]
+```
+
+`[1:]` means that we are skipping the first message. You can change, filter, and adapt the chat history to your use cases. `instance.serialize()` gives you the power to access all the chat logs. 
+
+## Mistral models 
+```python
+
+import os
+import panel as pn
+from mistralai.client import MistralClient
+from mistralai.models.chat_completion import ChatMessage
+
+pn.extension()
+
+
+async def callback(contents: str, user: str, instance: pn.chat.ChatInterface):
+    model = "mistral-small"
+    messages = [
+        ChatMessage(role=message["role"], content=message["content"])
+        for message in instance.serialize()[1:]
+    ]
+    print(messages)
+    response = client.chat_stream(model=model, messages=messages)
+
+    message = ""
+    for chunk in response:
+        part = chunk.choices[0].delta.content
+        if part is not None:
+            message += part
+            yield message
+
+
+client = MistralClient(api_key=os.environ["MISTRAL_API_KEY"])
+chat_interface = pn.chat.ChatInterface(callback=callback, callback_user="Mixtral")
+chat_interface.send(
+    "Send a message to get a reply from Mixtral!", user="System", respond=False
+)
+chat_interface.servable()
+```
+
+Here in this example, the model indeed knows what we were talking about previously. 
+
+<img src="./images/mistral_history.png" width="100%" style="margin-left: auto; margin-right: auto; display: block;"></img>
+
+## OpenAI models
+The code for OpenAI models is even simpler. Simply change `messages` to `instance.serialize()[1:]`, you will send all the chat history except for the first message to OpenAI API. 
+
+```python
+import panel as pn
+from openai import AsyncOpenAI
+
+pn.extension()
+
+
+async def callback(contents: str, user: str, instance: pn.chat.ChatInterface):
+    messages = instance.serialize()[1:]
+    response = await aclient.chat.completions.create(
+        model="gpt-3.5-turbo",
+        messages=messages,
+        stream=True,
+    )
+    message = ""
+    async for chunk in response:
+        part = chunk.choices[0].delta.content
+        if part is not None:
+            message += part
+            yield message
+
+
+aclient = AsyncOpenAI()
+chat_interface = pn.chat.ChatInterface(callback=callback, callback_user="ChatGPT")
+chat_interface.send(
+    "Send a message to get a reply from ChatGPT!", user="System", respond=False
+)
+chat_interface.servable()
+```
+
+<img src="./images/open_ai_history.gif" width="100%" style="margin-left: auto; margin-right: auto; display: block;"></img>
+
+
+# Adding download button to download all conversations 
+What if you'd like download all the conversations and chat histories? What if you would like to use your chat history for fine-tuning? You can add a download button to download all conversations. 
+
+
+<img src="./images/open_ai_download.gif" width="100%" style="margin-left: auto; margin-right: auto; display: block;"></img>
```

Comment:
If it's not too difficult, can we remake this GIF since the hearts icons are broken? But no big deal if not.